core/include/detray/builders/cylinder_portal_generator.hpp

```diff
@@ -0,0 +1,355 @@
+    scalar_t m_envelope{100.f * unit<scalar_t>::um};
+    /// Fixed inner radius during autofit
+    scalar_t m_fixed_inner_r{0.f};
+    /// Fixed outer radius during autofit
+    scalar_t m_fixed_outer_r{0.f};
+    /// Fixed length of the cylinder
+    scalar_t m_fixed_z{0.f};
+    /// The portal volumes links (north, south, east, west)
+    std::vector<dindex> m_volume_links{dindex_invalid, dindex_invalid,
+                                       dindex_invalid, dindex_invalid};
+
+    /// Setters
+    /// @{
+    constexpr cylinder_portal_config &build_inner(const bool b) {
+        m_build_inner = b;
+        return *this;
+    }
+    constexpr cylinder_portal_config &envelope(const scalar_t e) {
+        m_envelope = e;
+        return *this;
+    }
+    constexpr cylinder_portal_config &fixed_inner_radius(const scalar_t r) {
+        m_fixed_inner_r = r;
+        return *this;
+    }
+    constexpr cylinder_portal_config &fixed_outer_radius(const scalar_t r) {
+        m_fixed_outer_r = r;
+        return *this;
+    }
+    constexpr cylinder_portal_config &fixed_half_length(const scalar_t z) {
+        m_fixed_z = z;
+        return *this;
+    }
+    constexpr cylinder_portal_config &link_north(const dindex l) {
+        m_volume_links[0] = l;
+        return *this;
+    }
+    constexpr cylinder_portal_config &link_south(const dindex l) {
+        m_volume_links[1] = l;
+        return *this;
+    }
+    constexpr cylinder_portal_config &link_east(const dindex l) {
+        m_volume_links[2] = l;
+        return *this;
+    }
+    constexpr cylinder_portal_config &link_west(const dindex l) {
+        m_volume_links[3] = l;
+        return *this;
+    }
+    /// @}
+
+    /// Getters
+    /// @{
+    constexpr scalar_t envelope() const { return m_envelope; }
+    constexpr bool build_inner() const { return m_build_inner; }
+    constexpr scalar_t fixed_inner_radius() const { return m_fixed_inner_r; }
+    constexpr scalar_t fixed_outer_radius() const { return m_fixed_outer_r; }
+    constexpr scalar_t fixed_half_length() const { return m_fixed_z; }
+    constexpr const auto &volume_links() const { return m_volume_links; }
+    constexpr const auto &link_north() const { return m_volume_links[0]; }
+    constexpr const auto &link_south() const { return m_volume_links[1]; }
+    constexpr const auto &link_east() const { return m_volume_links[2]; }
+    constexpr const auto &link_west() const { return m_volume_links[3]; }
+    /// @}
+};
+
+/// @brief Generates a portal box around a volume that already contains surfaces
+///
+/// @tparam detector_t the type of detector the volume belongs to.
+template <typename detector_t>
+class cylinder_portal_generator final
+    : public surface_factory_interface<detector_t> {
+
+    using scalar_t = typename detector_t::scalar_type;
+    using point3_t = typename detector_t::point3;
+    using vector3_t = typename detector_t::vector3;
+    using transform3_t = typename detector_t::transform3;
```

Comment:
At one point we'll really need to unify the naming of the types in the repository. 🤔

My own preference is:
  - Call types in the `template` declaration `<foo>_t`. Exactly like how `detector_t` is treated in this example.
  - Declare `<foo>_type` typedefs for everything inside of the class/struct. Including the types that it directly receives as template arguments. I.e. in this case I'd add:

```c++
using detector_type = detector_t;
```

(This can be **very** useful when writing templated code on top of such classes.)

This is not for this PR, since there is just no uniform naming in any of the existing code. But in the non-too-distant future we should really do this. 🤔

For this PR: You may want to make the type declarations public. I don't see a good reason for why these should be private.

Reply:
Ok, I usually only put the ```_type``` ending on stuff that should be public. Anything internal to the class just keeps ```_t```, like the template parameters. But I agree, we need a LOT of cleanup here.

I was actually wondering how much of the type definitions should be public, because I don't like the case where client code gets the algebra information from some random other class that happens to be around just because it was convenient at the time of writing. This makes changing thing later on cumbersome, because it creates dependencies that are otherwise nonsense